### PR TITLE
feat: Validate swagger definitions

### LINKED ISSUE #3522: Autogenerated docs for REST API
Currently, you can examine [./codersdk](https://github.com/coder/coder/tree/main/codersdk) to see the various REST API queries supported by Coder. If you're interested in official documentation for the REST API (like [similar to our approach for Coder Classic](https://apidocs.coder.com)), please leave upvote this issue or [contact us](https://github.com/coder/coder). 

@mtojek:

Track progress

**OSS**

`/api/v2`:
- [x] `/` https://github.com/coder/coder/pull/5493
- [x] `/csp/reports` https://github.com/coder/coder/pull/5493
- [x] `/buildinfo` https://github.com/coder/coder/pull/5493
- [x] `/updatecheck` https://github.com/coder/coder/pull/5493
- [x] `/config/deployment` https://github.com/coder/coder/pull/5506
- [x] `/audit` and `/audit/testgenerate` https://github.com/coder/coder/pull/5506
- [x] `/files` https://github.com/coder/coder/pull/5506
- [x] `/parameters` https://github.com/coder/coder/pull/5506
- [x] `/authcheck` https://github.com/coder/coder/pull/5477
- [x] `/applications` https://github.com/coder/coder/pull/5477
- [x] `/workspaceagents` (~11 methods) https://github.com/coder/coder/pull/5538
- [x] `/workspacebuilds/{workspacebuild}` (~5 methods) https://github.com/coder/coder/pull/5538
- [x] `/workspaces` (~9 methods) https://github.com/coder/coder/pull/5383 https://github.com/coder/coder/pull/5538
- [x] `/templates/{template}` (~7 methods) https://github.com/coder/coder/pull/5383 https://github.com/coder/coder/pull/5546
- [x] `/templateversions/{templateversion}` (~11 methods) https://github.com/coder/coder/pull/5546
- [x] `/organizations` (~12 methods) https://github.com/coder/coder/pull/5383 https://github.com/coder/coder/pull/5546
- [x] `/users` (~23 methods) https://github.com/coder/coder/pull/5620

**Enterprise**

`/api/v2`: https://github.com/coder/coder/pull/5625
- [x] `/entitlements` (1 method)
- [x] `/replicas` (2 methods)
- [x] `/licenses` (4 methods)
- [x] `/organizations/{organization}/groups` (3 methods)
- [x] `/organizations/{organization}/provisionerdaemons` (2 methods)
- [x] `/groups/{group}` (3 methods)
- [x] `/workspace-quota` (1 method)
- [x] ~~`/service-banner`~~ `/appearance` (2 methods)
- [x] `/scim/v2` (6 methods)

**Validation**
- [ ] https://github.com/coder/coder/pull/5694

**UI improvements**
- [x] https://github.com/coder/coder/pull/5567 [Authentication](https://coder.com/docs/coder-oss/latest/api/authentication) page: explain how a user can generate tokens, link with [automation](https://coder.com/docs/coder-oss/latest/admin/automation#tokens) page

**Other**
- [x] https://github.com/coder/coder/pull/5590

### PR DESCRIPTION
**Pull request by mtojek**
Fixes: https://github.com/coder/coder/issues/3522

This PR concludes the autogenerating docs initiative with a basic Swagger validator and custom rules, that will let us keep the better quality of docs. The validator uses `go/ast` to parse comments and Go structure fields. It covers OSS and Enterprise `chi` handlers.

Checks:
1. Ensure consistency between `@Summary` and `@ID`.
2. Verify `@Success` and/or `@Failure`.
3. Verify presence of required annotations.
4. Make sure that Go comment (above the function) is before swagger's `@...`
5. Ensure that path parameters are defined.
6. Ensure that `@Security` tag is present.
7. Verify `@Accept` annotation.
8. Verify `@Produce` annotation.
9. Ensure unique swagger routes - no copy/paste mistakes.
10. Make sure that `uuid.UUID` and `time.Time` fields are formatted.

To reviewer:
- There are many files in this PR, mainly due to minor fixes spotted while implementing the validator. I suggest starting the peer review with `coderdtest/swaggerparser.go`.

### CODE REVIEW

File: coderd/coderdtest/swaggerparser.go [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+				kind: args[2],
+			}
+			c.parameters = append(c.parameters, p)
+		} else if strings.HasPrefix(text, "@Accept ") {
+			args := strings.SplitN(text, " ", 2)
+			c.accept = args[1]
+		} else if strings.HasPrefix(text, "@Produce ") {
+			args := strings.SplitN(text, " ", 2)
+			c.produce = args[1]
+		}
```
Comment (johnstcn) [inline coderd/coderdtest/swaggerparser.go:155]: Would this be better refactored as a switch statement?

Also, what do we want to do in the case of a prefix that doesn't match any of the above?

**Review by johnstcn — COMMENTED** [review]
I haven't gone through the generated docs with a fine-toothed comb, but I just had a couple of questions/observations about the Swagger parser.

File: coderd/coderdtest/swaggerparser.go (hunk 1 above)
Comment (mtojek) [reply coderd/coderdtest/swaggerparser.go:155]: Sure, I can refactor it, maybe I can squeeze it a bit.

> Also, what do we want to do in the case of a prefix that doesn't match any of the above?

We ignore it as it means that's a property that the validator does not process.

Comment (johnstcn) [reply coderd/coderdtest/swaggerparser.go:155]: > We ignore it as it means that's a property that the validator does not process.
👍

**Review by johnstcn — APPROVED** [review]
This is pretty neat 👍

**Comment by Emyrk** [comment]
Love how easily the go parser was to get comment groups. 

I would suggest we should comment what each assertion does. Some are obvious, but it would be helpful to have some english.

Do you intend to parse the go code at all? Like enforce all routes have a swagger defintion? Maybe digging into the functions and seeing if the return type is called via `httapi.Write`?

**Pull request merged**